### PR TITLE
Servicehook namespace permissions

### LINKED ISSUE #504: Add support for ServiceHook permission namespace
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description
I would like to set permissions in de ServiceHooks namespace with terraform

### New or Affected Resource(s)

None

### Potential Terraform Configuration

```hcl
resource "azuredevops_servicehook_permissions" "servicehook_permissions" {
  project_id = var.project_id
  principal  = var.principal_id
  permissions = {
    VIEW_SUBSCRIPTIONS = "allow"
    EDIT_SUBSCRIPTIONS = "allow"
    DELETE_SUBSCRIPTIONS = "allow"
    PUBLISH_EVENTS = "allow"
  }
}
```

### PR DESCRIPTION
**Pull request by martenvd**
## All Submissions:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.
* [x] All new and existing tests passed.
* [x] My code follows the code style of this project.
* [x] I ran lint checks locally prior to submission.
* [x] Have you checked to ensure there aren't other open PRs for the same update/change?

## Explanation of the changes
This PR should resolve issue 504, for the `azuredevops_servicehooks_permissions` resource can manage all the allow and deny bits for the servicehooks security namespace in Azure DevOps. I would like you to include these changes for our users and engineers have requested them and it would make managing our current codebase (Azure landingzones) easier.

## What about the current behavior has changed?

Issue Number: 504
https://github.com/microsoft/terraform-provider-azuredevops/issues/504

## Does this introduce a change to `go.mod`, `go.sum` or `vendor/`?

- [ ] Yes
- [x] No

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: azuredevops/internal/service/permissions/resource_servicehooks_permissions.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+		Create: resourceServiceHooksPermissionsCreateOrUpdate,
+		Read:   resourceServiceHooksPermissionsRead,
+		Update: resourceServiceHooksPermissionsCreateOrUpdate,
+		Delete: resourceServiceHooksPermissionsDelete,
+		Schema: securityhelper.CreatePermissionResourceSchema(map[string]*schema.Schema{
+			"project_id": {
+				Type:         schema.TypeString,
+				ValidateFunc: validation.IsUUID,
+				Optional:     true,
+				Required:     false,
```
Comment (xuzhang3) [inline azuredevops/internal/service/permissions/resource_servicehooks_permissions.go]:
```suggestion
				Optional:     true,
```

File: website/azuredevops.erb [hunk 2]
```diff
@@ -214,6 +214,9 @@
                 <li>
                   <a href="/docs/providers/azuredevops/r/serviceendpoint_npm.html">azuredevops_serviceendpoint_npm</a>
                 </li>
+                <li>
+                  <a href="/docs/providers/azuredevops/r/servicehooks_permissions.html.markdown">azuredevops_servicehooks_permissions</a>
```
Comment (xuzhang3) [inline website/azuredevops.erb]:
```suggestion
                  <a href="/docs/providers/azuredevops/r/servicehooks_permissions.html">azuredevops_servicehooks_permissions</a>
```

File: website/docs/r/servicehooks_permissions.html.markdown [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  Manages permissions for AzureDevOps service hooks
+---
+
+# azuredevops_servicehooks_permissions
+
+Manages permissions for service hooks
+
+## Permission levels
+
+Permissions for service hooks within Azure DevOps can be applied only on Project level.
```
Comment (xuzhang3) [inline website/docs/r/servicehooks_permissions.html.markdown]: If my understanding is right, the permission can be assigned to organization level if `project_id` not set. ping @tmeckel   if I'm wrong

File: azuredevops/provider.go [hunk 4]
```diff
@@ -74,6 +74,7 @@ func Provider() *schema.Provider {
 			"azuredevops_team_members":                           core.ResourceTeamMembers(),
 			"azuredevops_team_administrators":                    core.ResourceTeamAdministrators(),
 			"azuredevops_serviceendpoint_permissions":            permissions.ResourceServiceEndpointPermissions(),
+			"azuredevops_servicehooks_permissions":               permissions.ResourceServiceHooksPermissions(),
```
Comment (xuzhang3) [inline azuredevops/provider.go]: I would suggest rename `azuredevops_servicehooks_permissions` to `azuredevops_servicehook_permissions`

File: azuredevops/provider_test.go [hunk 5]
```diff
@@ -62,6 +62,7 @@ func TestProvider_HasChildResources(t *testing.T) {
 		"azuredevops_team_members",
 		"azuredevops_team_administrators",
 		"azuredevops_serviceendpoint_permissions",
+		"azuredevops_servicehooks_permissions",
```
Comment (xuzhang3) [inline azuredevops/provider_test.go]: same above, rename to `azuredevops_servicehook_permissions`

**Review by xuzhang3 — CHANGES_REQUESTED** [review]
@martenvd Thanks for your contribute. I few change requested, others looks good to me.

File: website/docs/r/servicehooks_permissions.html.markdown (hunk 3 above)
Comment (tmeckel) [reply website/docs/r/servicehooks_permissions.html.markdown]: @xuzhang3 exactly you're right! Thus the `project_id` is optional. @martenvd you can copy (or compare) my documentation

File: azuredevops/provider.go (hunk 4 above)
Comment (martenvd) [reply azuredevops/provider.go]: Thanks for the comments @xuzhang3, I have resolved them 👍

**Comment by martenvd** [comment]
Thanks for the comments @xuzhang3, I have resolved them 👍

File: website/azuredevops.erb [hunk 6]
```diff
@@ -214,6 +214,9 @@
                 <li>
                   <a href="/docs/providers/azuredevops/r/serviceendpoint_npm.html">azuredevops_serviceendpoint_npm</a>
                 </li>
+                <li>
+                  <a href="/docs/providers/azuredevops/r/servicehooks_permissions.html">azuredevops_servicehooks_permissions</a>
```
Comment (xuzhang3) [inline website/azuredevops.erb]:
```suggestion
                  <a href="/docs/providers/azuredevops/r/servicehooks_permissions.html">azuredevops_servicehook_permissions</a>
```

File: website/docs/r/servicehooks_permissions.html.markdown [hunk 7]
```diff
@@ -0,0 +1,72 @@
+---
+layout: "azuredevops"
+page_title: "AzureDevops: azuredevops_servicehooks_permissions"
```
Comment (xuzhang3) [inline website/docs/r/servicehooks_permissions.html.markdown]:
```suggestion
page_title: "AzureDevops: azuredevops_servicehook_permissions"
```

File: website/docs/r/servicehooks_permissions.html.markdown [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+  visibility         = "private"
+  description        = "Managed by Terraform"
+}
+
+data "azuredevops_group" "project-readers" {
+  project_id = azuredevops_project.project.id
+  name       = "Readers"
+}
+
+resource "azuredevops_servicehooks_permissions" "root-permissions" {
```
Comment (xuzhang3) [inline website/docs/r/servicehooks_permissions.html.markdown]:
```suggestion
resource "azuredevops_servicehook_permissions" "root-permissions" {
```

File: azuredevops/internal/acceptancetests/resource_servicehooks_permissions_test.go [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+func hclServiceHooksPermissions(projectName string, permissions map[string]map[string]string) string {
+	rootPermissions := datahelper.JoinMap(permissions["root"], "=", "\n")
+
+	return fmt.Sprintf(`
+%s
+data "azuredevops_group" "tf-project-readers" {
+	project_id = azuredevops_project.project.id
+	name       = "Readers"
+}
+resource "azuredevops_servicehooks_permissions" "acctest" {
```
Comment (xuzhang3) [inline azuredevops/internal/acceptancetests/resource_servicehooks_permissions_test.go]:
```suggestion
resource "azuredevops_servicehook_permissions" "acctest" {
```

File: azuredevops/internal/acceptancetests/resource_servicehooks_permissions_test.go [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+	})
+	config2 := hclServiceHooksPermissions(projectName, map[string]map[string]string{
+		"root": {
+			"ViewSubscriptions":   "Deny",
+			"EditSubscriptions":   "Deny",
+			"DeleteSubscriptions": "NotSet",
+			"PublishEvents":       "Allow",
+		},
+	})
+	tfNodeRoot := "azuredevops_servicehooks_permissions.acctest"
```
Comment (xuzhang3) [inline azuredevops/internal/acceptancetests/resource_servicehooks_permissions_test.go]:
```suggestion
	tfNodeRoot := "azuredevops_servicehook_permissions.acctest"
```

File: azuredevops/internal/acceptancetests/resource_servicehooks_permissions_test.go [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+	projectName := testutils.GenerateResourceName()
+	config := hclServiceHooksPermissions(projectName, map[string]map[string]string{
+		"root": {
+			"ViewSubscriptions":   "Deny",
+			"EditSubscriptions":   "NotSet",
+			"DeleteSubscriptions": "Deny",
+			"PublishEvents":       "Deny",
+		},
+	})
+	tfNodeRoot := "azuredevops_servicehooks_permissions.acctest"
```
Comment (xuzhang3) [inline azuredevops/internal/acceptancetests/resource_servicehooks_permissions_test.go]:
```suggestion
	tfNodeRoot := "azuredevops_servicehook_permissions.acctest"
```

**Review by xuzhang3 — CHANGES_REQUESTED** [review]

**Comment by xuzhang3** [comment]
@martenvd   some of the resource names are not renamed to `azuredevops_servicehook_permissions`

**Comment by martenvd** [comment]
@xuzhang3 I renamed all to servicehook instead of servicehooks

**Comment by xuzhang3** [comment]
@martenvd LGTM

**Comment by martenvd** [comment]
@xuzhang3 @tmeckel I have one question, is it normal that I'm not included in the contributors for this repo? Haven't contributed to open source projects in the past, so I have no idea how it is supposed to work.

**Comment by tmeckel** [comment]
@martenvd  if you look in the author info of the [`resource_servicehook_permissions.go`](https://github.com/microsoft/terraform-provider-azuredevops/blob/main/azuredevops/internal/service/permissions/resource_servicehook_permissions.go) file you'll notice that GitHub isn't able to resolve your GitHub account. 

image

As a comparison you can check [`baseSchema.go`](https://github.com/microsoft/terraform-provider-azuredevops/blob/main/azuredevops/internal/service/permissions/utils/baseSchema.go) there are two accounts listed as contributors:

image

**Comment by martenvd** [comment]
@tmeckel thanks for clarifying that! Too bad that is the case, next contributions will be under the right username!

**Pull request merged**